protocol/contracts/oracle/src/state/supported_pairs.rs

```diff
@@ -212,6 +218,59 @@ where
     }
 }
 
+fn check_tree_tickers(
+    tree: &Tree,
+    base_currency: &SymbolSlice,
+    stable_currency: &SymbolSlice,
+) -> Result<(), ContractError> {
+    if tree.root().value().target != base_currency {
+        return Err(ContractError::InvalidBaseCurrency(
+            tree.root().value().target.clone(),
+            ToOwned::to_owned(base_currency),
+        ));
+    }
+
+    // check for duplicated nodes
+    let mut supported_currencies: Vec<_> = tree
+        .iter()
+        .map(|node| node.value().target.as_ref())
+        .collect();
+
+    supported_currencies.sort();
+
+    if supported_currencies
+        .binary_search(&stable_currency)
+        .is_err()
+    {
+        return Err(ContractError::StableCurrencyNotInTree {});
+    }
+
+    if (0..supported_currencies.len() - 1)
+        .any(|index| supported_currencies[index] == supported_currencies[index + 1])
+    {
+        return Err(ContractError::DuplicatedNodes {});
```

Comment:
let's follow the single exit function point idiom and refrain from exiting in the middle of the function. I'd suggest merging all if-statements.